Member service

honeyl3ee opened the pull request:
- join method 구현
- joinConfirm method 구현
- test 작성
- dto 작성
- TokenSender 인터페이스 및 구현체 작성
- confirm 클래스 작성

kshshkim committed:
Feat: Content 목록 조회(무한 스크롤) (#38)

* Chore: querydsl 의존성 추가

* Feat: HashtagContent에 createdAt 필드(컬럼) 추가

무한 스크롤 구현 & 쿼리 성능을 위해서 정규화를 약간 희생함.

* Test: 테스트용 Fixture 추가

* Feat: ContentQueryRepository 목록조회(무한 스크롤)

- ContentQueryParameter, ContentQueryResult 쿼리 파라미터 DTO, 쿼리 결과 래핑용 DTO

Ref: #34

honeyl3ee committed:
Init: Add TokenMismatchException

인증코드가 일치하지 않을 시 발생하는 에러를 추가함
Github issue #33

honeyl3ee committed:
Init: Add TokenMismatchException

인증코드가 일치하지 않을 시 발생하는 에러를 추가함
Github issue #33

honeyl3ee committed:
Init: Add confirm entity

가입 승인을 위한 엔티티를 추가함
Github issue #33

honeyl3ee committed:
Init: Add confirm repository

confirm repository 추가
Github issue #33

honeyl3ee committed:
Init: Add passwordhasherImpl

paasswordHasher 구현체 추가
Github issue #33

honeyl3ee committed:
Refactor: Remove confirm field

승인된 회원인지 확인하는 상태값은 member 테이블로 옮겨 매번 confirm 테이블을 참조하지 않게 하는 것이 바람직해보임
Github issue #33

jaekkang committed:
Feat: MemberRepository, Member 리팩터링, 테스트 작성 외 (#45)

* Init: Create member repository

멤버 데이터를 저장하기 위한 저장소를 만들었습니다.
기본적인 영속성 추가를 위한
save()와 단건 데이터 검색과 전체 조회 메서드를 추가했습니다.

* Refactor: Change MemberRepository type

MemberRepository의 형식을 JPA Repository를 상속받는 형식으로 수정하였습니다.

* Feat: Add method for findByName

name을 기준으로 검색을 하는 메서드를 추가했습니다.

* Refactor: Change logic in create method

생성 메서드를 Builder 사용하는 방식으로 수정하였습니다.

* Style: Change method name

잘못 작성한 메서드의 이름을 변경하였습니다.
findByName -> findByusername

* Test: Add MemberRepository Test

작성한 Member Repository Test 추가했습니다.

간단히
- 빌더통해 생성 확인
- 저장한 데이터 불러오기
- 잘못된 색인시 에러 나는지

* Feat: Add hasher

Hasher 부분이 사용되지 않은 것을 확인해서 추가했습니다.

* Feat: Remove validate

validator가 잘못된 곳에서 사용되고 있어서 삭제했습니다.

* Feat: Add optional for empty data

Repository에서 서치시 데이터가 없을 경우를 위해 Optional로 감쌋습니다.

* Test: remove default method

빌더를 통한 객체 생성 테스트를 제외하고 jpaRepository에서 자동 생성되는 메서드의 테스트는 삭제했습니다.

* Style: Change method name

- validatePassword를 의미적으로 좀 더 맞는
verifyPassword로 변경되었습니다.

- 메서드에 대한 설명도 추가되었습니다.

* Test: Change test

Test문이 줄어 빌더 생성관련 밖에 남지 않아 MemberTest로 변경되었습니다.

honeyl3ee committed:
Init: Add token sender

회원가입 요청시 토큰을 전송하는 메서드를 멤버서비스에서 분리함.
Github issue #33

honeyl3ee committed:
Init: Add memberServiceImpl

join 메서드 구현
joinConfrim은 member에 승인 상태값 필드 추가하고, 업데이트 하는 로직 추가 필요함
Github issue #33

honeyl3ee committed:
Test: Add memberServiceImpl test

가입 요청 시 member 객체가 올바르게 저장되는지 테스트함
가입 요청 시 confirm 객체에 6자리 승인코드가 제대로 저장되는지 테스트함
Github issue #33

honeyl3ee committed:
Feat: Add repository annotation

어노테이션 누락으로 추가함
Github issue #33

honeyl3ee committed:
Style: Change code style

코드 컨벤션 맞춤
Github issue #33

yhk313 committed:
Feat: ContentCommandService 구현 외 (#44)

* Feat: HashtagContentRepository 생성

* feat: ContentSpecificData 에 해시태그 추가 및 getHashtagsList 메서드 추가

ContentSpecificData 에 해시태그를 추가하였고 HashtagContent 에 getHashtagsList 추가하여 해시태그 문자열을 분할하였습니다.

* feat: ContentController 구현

* Chore: incrementViewCount, incrementShareCount 메소드 수정

카운트 null 방지

* Test: ContentCommandImplServiceImplTest 테스트 코드 작성

* Feat: Content 엔티티에 hashtags 추가 및 getHashtagsList 메소드 이동

* chore: 카운트 증가 메소드 수정, 예외처리 수정

* chore: content 엔티티 hashtags columDefinition 설정

honeyl3ee committed:
Feat: 회원 가입 승인 정보 추가 (#46)

* Init: Add confirm stauts

멤버의 승인 상태를 나타내는 enum 추가
Github issue #41

* Feat: Add confirm status field

승인 상태를 나타내는 string 타입의 enum field 추가
Github issue #41

* Feat: Add updateConfirmStatus method

승인 상태를 변경할 수 있는 업데이트 메서드 추가
Github issue #41

* Refactor: Rename updateConfirmStatus to joinConfirm

가입이 승인된 후에 승인 상태가 바뀔 이유가 없어서 업데이트가 아닌 승인으로만 변경해주는 메서드로 수정함.
Github issue #41

honeyl3ee committed:
Init: Add Token Sender

이메일 전송, 토큰 생성 기능을 하는 클래스 추가
Github issue #33

honeyl3ee committed:
Refactor: Change class type

token sender를 인터페이스와 구현체로 분리함
Github issue #33

honeyl3ee committed:
Init: Add ConfirmTokenDto

ConfirmToken controller와 joinConfirm service에서 사용할 dto 추가
Github issue #5

honeyl3ee committed:
Refactor: Change jonConfirm parameter

필드에서 dto를 받는 것으로 변경함
Github issue #33

honeyl3ee committed:
Feat: Implement joinConfirm

joinConfirm 구현
Github issue #33

honeyl3ee committed:
Test: Add joinConfirm test

가입 요청 시 승인 상태 변경, 계정 일치, 패스워드 일치, 토큰 일치 검증 테스트
Github issue #33

honeyl3ee committed:
Refactor: Rename CreateMemberDto

CreateMemberDto의 request, response 클래스의 이름을 명확하도록 변경함

honeyl3ee committed:
Init: Add password hasher implementation

hasher 구현체 작성

honeyl3ee committed:
Merge branch 'master' into memberService

honeyl3ee committed:
Style: Confirm

코드 컨벤션 적용

kshshkim commented on src/main/java/beforespring/socialfeed/member/domain/ConfirmRepository.java:
Optional로 감싸주는 것이 좋습니다.

kshshkim commented on src/main/java/beforespring/socialfeed/member/domain/PasswordHasherImpl.java:
필요 없는 클래스입니다. 제거해주세요.

kshshkim replied:
PasswordHasher는 구현이 된 상태입니다. 마스터 브랜치의 변경사항을 반영해주세요

kshshkim commented on src/main/java/beforespring/socialfeed/member/infra/TokenSenderImpl.java:
System.out.println은 바람직하지 못한 로깅 방식입니다.

`@Slf4j`에 대해 검색해보시고, 로깅 방법을 바꿔주세요.

honeyl3ee replied:
이건 로깅을 위해서 출력을 한 건 아니었고 이메일을 보내는 기능이 생략되어 있어서 임의로 넣어둔 출력문입니다.

honeyl3ee replied:
수정 완료했습니다

kshshkim commented on src/main/java/beforespring/socialfeed/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+        //추후 주입 받는 것으로 변경
+        PasswordHasher passwordHasher = new PasswordHasherImpl();
+        if (!passwordHasher.matches(request.getPassword(), member.getPassword()))
+            throw new PasswordMismatchException();
```
Member 클래스에 비밀번호 검증 메서드가 존재합니다. Member 객체의 verifyPassword 메서드를 이용해주세요.

kshshkim commented on src/main/java/beforespring/socialfeed/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+    @Override
+    public void joinConfirm(ConfirmTokenRequest request) {
+        Member member = memberRepository.findByUsername(request.getUsername()).orElseThrow(IllegalAccessError::new);
+        Confirm confirm = confirmRepository.findByMember(member);
```
Confirm 클래스는 Member 클래스를 참조하고 있습니다. 두 번의 쿼리를 넣어서 따로 불러올 필요 없이, 객체의 필드를 통해 접근하면 됩니다.

kshshkim replied:
코드를 읽다보니 왜 이렇게 됐는지 알았습니다. Member 클래스에 양방향 연관관계를 설정하거나, 아니면 연관관계의 주인을 Member 클래스로 설정해도 될 것 같습니다.

honeyl3ee replied:
근데 그렇게 되면 member class에서 token 필드를 가지고 있는게 낫지 않나요? member class에서 token이 매번 필요한 필드가 아니라서 member id를 외래키로 가지고 token을 저장하는 confirm class를 생성했습니다.

jaekkang replied:
🤔

kshshkim replied:
memberId로 Confirm을 find하는 것이 맞아보이는데요, 이전 세션에서 멘토님도 말씀하셨듯이 이미 계정 정보는 생성이 된 상태이니, 회원ID와 인증토큰만 받아서 처리해도 될 것 같습니다.

kshshkim replied:
다시 생각해보니 Member 클래스에서 Confirm을 참조하지 않는 원래 결정이 맞아보입니다. 

memberId로 Confirm을 find하는 것이 맞아보이는데요, 이전 세션에서 멘토님도 말씀하셨듯이 이미 계정 정보는 생성이 된 상태이니, 메서드 자체에서 회원ID와 인증토큰만 받아서 처리해도 될 것 같습니다. 

그런데 이 경우는 쿼리를 따로 작성을 해주셔야합니다.

kshshkim commented on src/main/java/beforespring/socialfeed/member/domain/Confirm.java:
```diff
@@ -0,0 +1,45 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "memberId")
```
DB 컬럼 이름으로는 snake_case 혹은 SCREAMING_SNAKE_CASE를 사용하는 것이 표준적입니다.

kshshkim commented on src/main/java/beforespring/socialfeed/member/domain/Confirm.java:
RDB에서는 camelCase나 PascalCase를 잘 사용하지 않습니다. 

db의 테이블 이름이나 컬럼 이름은 snake_case 혹은 SCREAMING_SNAKE_CASE로 설정하는 것이 관례입니다.

kshshkim commented on src/main/java/beforespring/socialfeed/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+            throw new PasswordMismatchException();
+        if (!confirm.getToken().equals(request.getToken()))
+            throw new TokenMismatchException();
+        member.joinConfirm();
```
Member 클래스에는 이미 verifyPassword() 메서드가 존재합니다. 로직이 중복 구현된 상황입니다.
Member 클래스의 verifyPassword() 메서드를 참조하셔서, Confirm 클래스에도 유사한 메서드를 만드는 것이 좋아보입니다.

현재 구현된 메서드의 책임은 다음과 같습니다.
- 트랜잭션 설정
- 리포지토리에서 Member 엔티티 로드하기
- 리포지토리에서 Confirm 엔티티 로드하기
- passwordHasher의 matches 메서드 호출
- 비밀번호 일치 여부에 따른 분기 처리
- 토큰 일치 여부에 따른 분기 처리
- Member 객체의 joinConfirm() 호출

하나의 메서드에 책임이 너무 많아지는 문제가 있습니다.

kshshkim commented on src/main/java/beforespring/socialfeed/member/service/MemberServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+                            .build();
+        Long memberId = memberRepository.save(member).getId();
+        String token = tokenSender.generateToken();
+        tokenSender.sendEmail(request.getEmail(), token);
```
이 부분은 트랜잭션과 별개로, 트랜잭션이 성공한 후에 실행되도록 하는 것이 좋아보입니다. 지금은 더미 구현체이기 때문에 큰 문제가 없으나, 비동기로 이메일을 보내는 역할을 실제로 수행할 경우엔 문제가 발생할 수 있습니다.

confirmRepository.save()가 실패하더라도 메일이 전송되게 되는데, 이 경우 사용자는 계속해서 인증에 실패하게 됩니다.

kshshkim commented on src/main/java/beforespring/socialfeed/member/infra/TokenSenderImpl.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public String generateToken() {
+        StringBuilder code = new StringBuilder();
+        Random random = new Random();
```
random을 매번 새로 생성하는 것 보다, final 필드로 가지고 있는 편이 좋습니다.

kshshkim commented:
우선 master branch의 변경 사항 반영을 해주셔야할 것 같습니다.

honeyl3ee committed:
Merge branch 'master' into memberService

honeyl3ee committed:
Refactor: findByMember returns Optional

findByMember가 Optional<Confirm>을 반환하도록 변경함

honeyl3ee committed:
Init: Add exception

member와 confirm 조회 시 존재하지 않을 때 발생하는 예외 추가

honeyl3ee committed:
Refactor: sendEmail log

터미널 출력을 제거하고 적절한 로깅 방식 적용

honeyl3ee committed:
Remove: PasswordHasherImpl

불필요한 구현체 삭제

honeyl3ee committed:
Feat: rename field name, Add vefiyToken

camelCase를 snake_case로 변경
token 검증 메서드 추가

honeyl3ee committed:
Feat: Add save member, save confirm

db와 관련없는 로직은 트랜잭션에서 분리하여 구현함
joinConfirm의 토큰 검증 코드와 패스워드 검증은 별도의 클래스에서 진행되도록 책임을 줄임.

honeyl3ee committed:
Test: custom exception, change wrong parameter

custom exception 추가, confrimTokenRequest에 잘못된 인자를 수정함
